File: packages/julia/src/Seguranca/Ataques/AtaqueSeparacaoChaveIV.jl
"""
Testa se o keystream depende de key e iv APENAS pela combinação key XOR iv.
Nessa cifra o estado inicial é `a[pos] = key[pos] ^ iv[pos]`, então, se o resto
do gerador não reintroduzir key/iv separadamente, vale exatamente:

    keystream(key, iv) == keystream(key ^ d, iv ^ d)

para qualquer máscara d (com o devido alinhamento key/iv). Isso é uma
propriedade estrutural relevante: significa que key e iv não entram de forma
independente na cifra, o que enfraquece o modelo de segurança (o IV deveria
contribuir com entropia própria, não só deslocar a chave por XOR).
"""
struct AtaqueSeparacaoChaveIV <: Ataque
    tentativas::Int
    tamanho_bloco::Int
end

AtaqueSeparacaoChaveIV(; tentativas::Integer = 50, tamanho_bloco::Integer = 32) =
    AtaqueSeparacaoChaveIV(Int(tentativas), Int(tamanho_bloco))

nome(::AtaqueSeparacaoChaveIV) = "Separação chave/IV (invariância a key XOR iv)"

function executar(a::AtaqueSeparacaoChaveIV, alvo::AlvoCriptografico)
    tam_chave = length(codeunits(chave_de_teste(alvo)))
    tam_iv = tamanho_iv(alvo)

    confirmacoes = 0
    for _ in 1:(a.tentativas)
        chave = random_bytes(tam_chave)
        iv = random_bytes(tam_iv)
        d = random_bytes(tam_iv)

        chave2 = Vector{UInt8}(undef, tam_chave)
        for p in 0:(tam_chave-1)
            chave2[p+1] = chave[p+1] ⊻ d[mod(p, tam_iv)+1]
        end
        iv2 = Vector{UInt8}(undef, tam_iv)
        for p in 0:(tam_iv-1)
            iv2[p+1] = iv[p+1] ⊻ d[p+1]
        end

        ks1 = gerar_keystream_bruto(alvo, chave, iv, "enc", a.tamanho_bloco)
        ks2 = gerar_keystream_bruto(alvo, chave2, iv2, "enc", a.tamanho_bloco)

        if ks1 == ks2
            confirmacoes += 1
        end
    end

    invariante = confirmacoes == a.tentativas

    return ResultadoAtaque(
        nome(a),
        invariante,
        invariante ? "media" : "info",
        invariante ?
        "Confirmado em $(a.tentativas)/$(a.tentativas): keystream(key,iv) == keystream(key^d, iv^d). " *
        "O IV só desloca a chave por XOR antes da difusão - não injeta entropia independente no key schedule." :
        "Invariância não se confirmou ($confirmacoes/$(a.tentativas)); key e iv entram de forma independente.",
        Dict{String,Any}("confirmacoes" => confirmacoes, "tentativas" => a.tentativas),
    )
end

File: packages/java/src/main/java/fbc/Seguranca/Ataques/AtaqueSeparacaoChaveIV.java
package fbc.Seguranca.Ataques;

import fbc.Seguranca.AlvoCriptografico;
import fbc.Seguranca.AtaqueInterface;
import fbc.Seguranca.ResultadoAtaque;
import fbc.Seguranca.Severidade;
import fbc.Seguranca.SkipAtaqueException;
import fbc.Seguranca.Util;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * Testa se o keystream depende de key e iv APENAS pela combinação key XOR iv.
 * Nessa cifra o estado inicial é `a[pos] = key[pos] ^ iv[pos]`, então, se o
 * resto do gerador não reintroduzir key/iv separadamente, vale exatamente:
 *
 *     keystream(key, iv) == keystream(key ^ d, iv ^ d)
 *
 * para qualquer máscara d (com o devido alinhamento key/iv). Isso é uma
 * propriedade estrutural relevante: significa que key e iv não entram de
 * forma independente na cifra, o que enfraquece o modelo de segurança (o IV
 * deveria contribuir com entropia própria, não só deslocar a chave por XOR).
 */
public class AtaqueSeparacaoChaveIV implements AtaqueInterface {
    private final int tentativas;
    private final int tamanhoBloco;

    public AtaqueSeparacaoChaveIV() {
        this(50, 32);
    }

    public AtaqueSeparacaoChaveIV(int tentativas, int tamanhoBloco) {
        this.tentativas = tentativas;
        this.tamanhoBloco = tamanhoBloco;
    }

    @Override
    public String nome() {
        return "Separação chave/IV (invariância a key XOR iv)";
    }

    @Override
    public ResultadoAtaque executar(AlvoCriptografico alvo) {
        int tamChave = alvo.chaveDeTeste().getBytes(StandardCharsets.UTF_8).length;
        int tamIv = alvo.tamanhoIv();

        byte[] primeiro = alvo.gerarKeystreamBruto(
                alvo.chaveDeTeste().getBytes(StandardCharsets.UTF_8),
                Util.randomBytes(tamIv),
                "enc",
                this.tamanhoBloco);
        if (primeiro == null) {
            throw new SkipAtaqueException("Alvo não expõe gerarKeystreamBruto.");
        }

        int confirmacoes = 0;
        for (int t = 0; t < this.tentativas; t++) {
            byte[] chave = Util.randomBytes(tamChave);
            byte[] iv = Util.randomBytes(tamIv);
            byte[] d = Util.randomBytes(tamIv);

            byte[] chave2 = new byte[tamChave];
            for (int p = 0; p < tamChave; p++) {
                chave2[p] = (byte) ((chave[p] ^ d[p % tamIv]) & 0xFF);
            }
            byte[] iv2 = new byte[tamIv];
            for (int p = 0; p < tamIv; p++) {
                iv2[p] = (byte) ((iv[p] ^ d[p]) & 0xFF);
            }

            byte[] ks1 = alvo.gerarKeystreamBruto(chave, iv, "enc", this.tamanhoBloco);
            byte[] ks2 = alvo.gerarKeystreamBruto(chave2, iv2, "enc", this.tamanhoBloco);

            if (Arrays.equals(ks1, ks2)) {
                confirmacoes++;
            }
        }

        boolean invariante = confirmacoes == this.tentativas;

        Map<String, Object> dados = new HashMap<>();
        dados.put("confirmacoes", confirmacoes);
        dados.put("tentativas", this.tentativas);

        return new ResultadoAtaque(
                nome(),
                invariante,
                invariante ? Severidade.MEDIA : Severidade.INFO,
                invariante
                        ? "Confirmado em " + this.tentativas + "/" + this.tentativas
                                + ": keystream(key,iv) == keystream(key^d, iv^d). "
                                + "O IV só desloca a chave por XOR antes da difusão - não injeta entropia independente no key schedule."
                        : "Invariância não se confirmou (" + confirmacoes + "/" + this.tentativas
                                + "); key e iv entram de forma independente.",
                dados);
    }
}
